lib/open_food_network/variant_stock.rb

```diff
@@ -1,80 +1,136 @@
-# was previously a database column on variants.
+# These methods were available in Spree 1, but were removed in Spree 2.  We
+# would still like to use them so that we still give support to the consumers
+# of this methods, making the upgrade backward compatible.
+#
+# Therefore we use only a single stock item per variant, which is associated to
+# a single stock location per instance (default stock location) and use it to
+# track the `count_on_hand` value that was previously a database column on
+# variants. See
+# https://github.com/openfoodfoundation/openfoodnetwork/wiki/Spree-Upgrade%3A-Stock-locations
+# for details.
 #
 # We may decide to deprecate these methods after we designed the Network feature.
 module OpenFoodNetwork
   module VariantStock
     extend ActiveSupport::Concern
 
     included do
-      after_save :save_stock
+      after_update :save_stock
     end
 
+    # Returns the number of items of the variant available in the
+    # stock. When allowing on demand, it returns infinite.
+    #
+    # Note this method relies on Spree's 2.0 #total_on_hand.
+    #
+    # @return [Float|Integer]
     def on_hand
+      warn_deprecation(__method__)
+
       if on_demand
         Float::INFINITY
       else
         total_on_hand
       end
     end
 
+    # Alias for Spree's 2.0 #total_on_hand
+    #
+    # @return [Float|Integer]
     def count_on_hand
+      warn_deprecation(__method__)
       total_on_hand
     end
 
+    # Sets the stock level when `track_inventory_levels` config is
+    # set. It raises otherwise.
+    #
+    # @raise [StandardError] when the track_inventory_levels config
+    # key is not set.
     def on_hand=(new_level)
+      warn_deprecation(__method__)
+
       error = 'Cannot set on_hand value when Spree::Config[:track_inventory_levels] is false'
       raise error unless Spree::Config.track_inventory_levels
 
       self.count_on_hand = new_level
     end
 
+    # Sets the stock level. As opposed to #on_hand= it does not check `track_inventory_levels`'s value. It does ensure there's a stock item for the variant however raising otherwise. See #raise_error_if_no_stock_item_available for details.
+    #
+    # @raise [StandardError] when the variant has no stock item yet
     def count_on_hand=(new_level)
+      warn_deprecation(__method__)
+
       raise_error_if_no_stock_item_available
       overwrite_stock_levels new_level
     end
 
     def on_demand
-      stock_items.any?(&:backorderable?)
+      warn_deprecation(__method__)
+      stock_items.first.backorderable?
     end
 
+    # Sets whether the variant can be ordered on demand or not. Note
+    # that although this modifies the stock item, it is not
+    # persisted.
+    #
+    # @raise [StandardError] when the variant has no stock item yet
     def on_demand=(new_value)
+      warn_deprecation(__method__)
+
       raise_error_if_no_stock_item_available
+      raise_error_if_multiple_stock_items
```

Comment:
I'll try to convince you this deserves to become a DB-level constraint although opinions (See https://openfoodnetwork.slack.com/threads/convo/C4NDJT3FY-1535723508.000100/) were against it.

First, at this point, it is too late. The data is already inconsistent violating data integrity and we can cleaning it manually is the most we can do. And that is hard.

Second, nothing stops us data to become inconsistent by any other means. Adding this sort of barrier, doesn't prevent this from happening from a rails console, a controller or even a psql session. It's like *poner puertas al campo* aka. trying to stop the unstoppable.

As a result, unconfident programming comes up; the lines of code consuming `stock_items` become paranoid checking there's a single stock item throughtout the app.

That is why, I believe this would be better off replaced by `add_index :spree_stock_items, :variant_id, unique: true` in a migration. I know we don't like messing with Spree's schema but this doesn't modify the tables and the data integrity it provides well outweighs it.

Reply:
I see. I dont like the approach because we should know who's doing what and fix when it's not happening. 

"become paranoid checking there's a single stock item throughout the app"
The place from where code is "consuming stock_items" should be one and only one: it's the model and service layer of this part of the app. But I know we don't have that kind of isolation... So, lets do it!

Reply:
Yes, I'm in favour of a constraint. That does make some of my earlier comments redundant.